### ISSUE TITLE
Distance-based discrete interpolation in affine transformation

### ISSUE THREAD

**Issue by Spenhouet** [scikit-image/scikit-image#4824]
We perform affine transformations on 3D segmentation images.
An important requirement of this is that the interpolated values need to be discrete and consist only of existing classes.

With scipy the only way is to set the order to 0 to archive a nearest neighbor interpolation:

    scipy.ndimage.affine_transform(..., order=0)

This will result in a poor quality result but is very fast.

What works way better is a distance-based interpolation.
For every voxel a window is defined.
For all values within this window the spatial distance to the interpolation value is calculated.
The value (class) with the lowest total distance within this window will be the selected / resulting interpolation value.

We implemented our own version of that but the performance is abysmal.
Initially we thought that this is a very common use-case / functionality and did expect this to be implemented in most packages.
We couldn't find anything similar.

It would be great if scipy or scikit would have such a functionality. 

This could be implemented in a more generalized form allowing to specify the window size and if the final window computation should be discrete or linear.

**Comment by jni** [comment]
@Spenhouet thanks for raising this issue!

It sounds useful but I must admit I don't fully follow the reasoning. What do you mean by saying that ndimage gives a "poor quality result"? What is "quality" here? Could you show some examples with data?

Also, you mention you have a working but slow implementation — could you link to it? A baseline implementation is infinitely useful when trying to develop a corresponding fast one!

Other than that this functionality is definitely in-scope for scikit-image.

**Comment by Spenhouet** [comment]
> It sounds useful but I must admit I don't fully follow the reasoning. What do you mean by saying that ndimage gives a "poor quality result"? What is "quality" here? Could you show some examples with data?

Including multiple values around the interpolation value does result in more realistic interpolations. We are using this on medical images and using nearest neighbor does result in a loss of information aka poor quality.

> Also, you mention you have a working but slow implementation — could you link to it

I would like to but this is in our private source. For now I will not be able to do that. Sorry :/

**Comment by rfezzani** [comment]
> >    It sounds useful but I must admit I don't fully follow the reasoning. What do you mean by saying that ndimage gives a "poor quality result"? What is "quality" here? Could you show some examples with data?

> Including multiple values around the interpolation value does result in more realistic interpolations. We are using this on medical images and using nearest neighbor does result in a loss of information aka poor quality.

@Spenhouet can you please provide an illustration of what you consider good and bad interpolation. I am sorry but it's not really clear to me how "Including multiple values around the interpolation value does result in more realistic interpolations".

**Comment by Spenhouet** [comment]
> > Including multiple values around the interpolation value does result in more realistic interpolations. We are using this on medical images and using nearest neighbor does result in a loss of information aka poor quality.
> 
> @Spenhouet can you please provide an illustration of what you consider good and bad interpolation. I am sorry but it's not really clear to me how "Including multiple values around the interpolation value does result in more realistic interpolations".

Think of classification / segmentation. Just looking for the nearest neighbor class is not good enough. You will get much more accurate interpolations if you take more classifications in the area into account.

From my perspective this is not something that you can show visually but that makes sense if explained. Maybe I do a terrible job at explaining.

Lets look at an artificial example. The following is a 5x5 image with two classes annotated with "0" and "1":

> 0 0 0 0 0
> 0 0 1 0 0
> 0 0 ? 0 0
> 0 0 0 0 0
> 0 0 0 0 0

There is only one classification of the class "1". Lets assume this class is the nearest neighbor of the interpolation value "?".
Would this be a good interpolation of this value?
We probably would expect "0" as interpolation.

This is probably not the best example and there are many different cases but I hope the idea is clear.

It is about the majority. The interpolation should be the value of majority in the window. The majority is a more true representation of the data.

**Comment by rfezzani** [comment]
As you pointed out, the nearest neighbor is maybe not the good paradigm in your case.
What about applying [majority filter](https://scikit-image.org/docs/dev/api/skimage.filters.rank.html?highlight=otsu%20filter#skimage.filters.rank.majority) before affine transformation?

**Comment by Spenhouet** [comment]
@rfezzani That is why I mentioned that my example is only for one case. Using a majority filter would just solve this one case. All interpolation values should be based on the majority. Also as defined above the majority should be defined by the distance and not just the majority.

**Comment by rfezzani** [comment]
@Spenhouet, I am somehow skeptical about the approach that you are proposing :/
How can we guaranty that a small label area will not disappear with the majority criteria that you are proposing?
Your approach has also the drawback of modifying the input labels when the affine transformation is the identity...

**Comment by Spenhouet** [comment]
> How can we guaranty that a small label area will not disappear with the majority criteria that you are proposing?

I don't think that is an issue. That would be an expected outcome and depends on the window size.

> Your approach has also the drawback of modifying the input labels when the affine transformation is the identity...

Our approach yes and that is a valid point.
I don't know how the interpolation for the affine transformations is currently implemented (I think that is in some C / C++ file that I could not find).
I don't know if the current implementation performs the interpolation only for the points where one is necessary or for all.
I guess that this is done for all points.
Therefore the distance weighting would be applied to all values and therefore, as you point out, for points where that might not be necessary.

Nonetheless, I'm not sure in case of the identity and other trivial transformations that there is any interpolation performed at all. I would suspect that there is some shortcut implemented for these cases.


Continuing on your suggestion with the majority filter:
I wonder what the result of performing an affine transformation and then running a majority filter would look like and how close that would be (this would not be a solution, just thinking out load here)?
But even if these were two consecutive operations, it still would need to be distance based and work on n-dimensional images (at least 3D).

**Comment by jni** [comment]
@Spenhouet note that there is a 3D rank filters PR open here: https://github.com/scikit-image/scikit-image/pull/4435 You might be able to grab that code and use it for the two step approach.

Regarding the warping code, it's done [here](https://github.com/scikit-image/scikit-image/blob/3a4dc584d062347d9ab6bcd0910aeb7b40bd9623/skimage/transform/_warps_cy.pyx). Currently, it's 2D only. For 3D+, we actually wrap `scipy.ndimage.map_coordinates`, see [here](https://github.com/scikit-image/scikit-image/blob/3a4dc584d062347d9ab6bcd0910aeb7b40bd9623/skimage/transform/_warps.py#L927).

It's a longstanding dream of mine to get our warping to 3D/nD, see #3544 for a start with the transform classes, but it's a big task and a prerequisite for us to provide what you're after directly. In the meantime, a distance-weighted majority vote would be doable with `ndimage.generic_filter` and this can be fast-ish with Numba, see my blog posts on this topic [here](https://ilovesymposia.com/2017/03/12/scipys-new-lowlevelcallable-is-a-game-changer/) and [here](https://ilovesymposia.com/2017/03/15/prettier-lowlevelcallables-with-numba-jit-and-decorators/). So you can do the 2-step approach with SciPy.

If you can find a way within your company to share just the code for this functionality with a suitable (MIT/BSD/Apache) license, or even as a WIP/draft pull request that we could iterate on, that would be tremendously valuable to move forward. I estimate it's about 10-100x easier to speed up a reference implementation than to build one from scratch.
